Mathlib/RingTheory/TwoSidedIdeal/Matrix.lean

```diff
@@ -0,0 +1,153 @@
+
+In this file, we show that the two-sided ideals of `A` corresponds bijectively to that of `Mₙ(A)`.
+
+## Main results
+
+* `TwoSidedIdeal.equivRingConMatrix` : The equivalence between the two-sided ideals of `A` and
+  `Mₙ(A)` where sends `I ≤ A` to `Mₙ(I)` and `J ≤ Mₙ(A)` to `{x₀₀ | x ∈ J}`.
+* `TwoSidedIdeal.OrderEquivMatrix` : The order isomorphism between the two-sided ideals of `A`
+  and `Mₙ(A)`.
+-/
+
+variable (A : Type*) [Ring A]
+
+open BigOperators Matrix MulOpposite
+
+local notation "M[" ι "," R "]" => Matrix ι ι R
+
+namespace TwoSidedIdeal
+
+variable (ι : Type) [Fintype ι]
+
+/--
+If `I` is a two-sided-ideal of `A`, then `Mₙ(I) := {(xᵢⱼ) | ∀ i j, xᵢⱼ ∈ I}` is a two-sided-ideal of
+`Mₙ(A)`.
+-/
+@[simps]
+def mapMatrix (I : TwoSidedIdeal A) : TwoSidedIdeal M[ι, A] := .mk
+{
+  r := fun X Y => ∀ i j, I.ringCon (X i j) (Y i j)
+  iseqv :=
+  { refl := fun X i j ↦ I.ringCon.refl (X i j)
+    symm := fun h i j ↦ I.ringCon.symm (h i j)
+    trans := fun h1 h2 i j ↦ I.ringCon.trans (h1 i j) (h2 i j) }
+  mul' := by
+    intro _ _ _ _ h h' i j
+    rw [Matrix.mul_apply, Matrix.mul_apply]
+    rw [TwoSidedIdeal.rel_iff, ← Finset.sum_sub_distrib]
+    apply I.finsetSum_mem
+    rintro k -
+    rw [← TwoSidedIdeal.rel_iff]
+    apply I.ringCon.mul (h _ _) (h' _ _)
+  add' := fun {X X' Y Y'} h h' i j ↦ by
+    simpa only [Matrix.add_apply] using I.ringCon.add (h _ _) (h' _ _)
+}
+
+@[simp] lemma mem_mapMatrix (I : TwoSidedIdeal A) (x) : x ∈ I.mapMatrix A ι ↔
+    ∀ i j, x i j ∈ I :=
+  Iff.rfl
+
+variable [Inhabited ι]
+
+/-- If `I` is an two-sided ideal of `Mₙ(A)`, then `{M₀₀| M ∈ I}` is a two-sided ideal of `A`. -/
+@[simps!]
+def fromMatrix (I : TwoSidedIdeal M[ι, A]) :
+    TwoSidedIdeal A :=
+  .mk'
+    ((fun (x : M[ι, A]) ↦ x default default) '' I)
+    ⟨0, I.zero_mem, rfl⟩
+    (by rintro _ _ ⟨x, hx, rfl⟩ ⟨y, hy, rfl⟩; exact ⟨x + y, I.add_mem hx hy, rfl⟩)
+    (by rintro _ ⟨x, hx, rfl⟩; exact ⟨-x, I.neg_mem hx, rfl⟩)
+    (by
+      classical
+      rintro x _ ⟨y, hy, rfl⟩
+      exact ⟨Matrix.diagonal (fun _ ↦ x) * y, I.mul_mem_left _ _ hy, by simp⟩)
+    (by
+      classical
+      rintro _ y ⟨x, hx, rfl⟩
+      exact ⟨x * Matrix.diagonal (fun _ ↦ y), I.mul_mem_right _ _ hx, by simp⟩)
+
+variable [DecidableEq ι]
+
+/--
+The two-sided-ideals of `A` corresponds bijectively to that of `Mₙ(A)`.
+Given an ideal `I ≤ A`, we send it to `Mₙ(I)`.
+Given an ideal `J ≤ Mₙ(A)`, we send it to `{x₀₀ | x ∈ J}`.
+-/
+@[simps]
+def equivRingConMatrix : TwoSidedIdeal A ≃ TwoSidedIdeal M[ι, A] where
+  toFun I := I.mapMatrix A ι
+  invFun J := J.fromMatrix A
+  right_inv J := SetLike.ext fun x ↦ by
+    simp only [mem_mapMatrix]
+    constructor
+    · intro h
+      choose y hy1 hy2 using h
+      rw [Matrix.matrix_eq_sum_stdBasisMatrix x]
+      refine J.finsetSum_mem _ _ fun i _ ↦ J.finsetSum_mem _ _ fun j _ ↦ ?_
+      suffices
+          stdBasisMatrix i j (x i j) =
+          stdBasisMatrix i default 1 * y i j * stdBasisMatrix default j 1 by
```

Comment:
@Whysoserioushah, regarding #23641, this still needed unitality in the proof here.